modules/frontend/searchsharding.go

```diff
@@ -283,12 +197,22 @@ func (s searchSharder) RoundTrip(r *http.Request) (*http.Response, error) {
 			overallResponse.addResponse(results)
 		}(req)
 	}
+
+	// wait for all goroutines running in wg to finish or cancelled
 	wg.Wait()
 
+	// print out request metrics
+	cancelledReqs := startedReqs - overallResponse.finishedRequests
+	_ = level.Info(s.logger).Log(fmt.Sprintf(
+		"sharded search request stats, raw_query: %s, total: %d, started: %d, finished: %d, cancelled: %d",
+		r.URL.RawQuery, len(reqs), startedReqs, overallResponse.finishedRequests, cancelledReqs))
+
 	// all goroutines have finished, we can safely access searchResults fields directly now
 	span.SetTag("inspectedBlocks", overallResponse.resultsMetrics.InspectedBlocks)
```

Comment:
We are accessing the fields of overallResponse here, but there are still goroutines out that can modify it. We can either have the searchResponse object stop taking updates once `internalShouldQuit()` returns true. We can add accessors with appropriate locking for the fields that we need. This already partially exists with `result()`.

Reply:
wups! nevermind. I forgot we stuck with the `wg.Wait()` style. I was thinking about the channel selection we had discussed. this is good!